### PR TITLE
health_check: support sending headers with HTTP health check

### LINKED ISSUE #2888: Support sending headers with HTTP health checks
*Description*:
Thanks to some legacy cruft, my health check endpoint will return a 400 error if it is invoked without a specific set of HTTP headers.

This is an enhancement request to allow me to specify one or more headers to be included when Envoy calls the health check endpoint.

*Relevant Links*:
https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/health_check.proto.html#envoy-api-msg-core-healthcheck-httphealthcheck

### PR DESCRIPTION
**Pull request by dio**
health_check: support sending headers with HTTP health check

*Description*:
This patch adds support for sending headers with HTTP health check.

*Risk Level*: Low

*Testing*:
- Unit test on configuring additional headers

*Docs Changes*:
- https://github.com/envoyproxy/data-plane-api/pull/595

*Release Notes*:
- https://github.com/envoyproxy/data-plane-api/pull/595

Fixes #2888 

*API Changes*:
- https://github.com/envoyproxy/data-plane-api/pull/595

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

### CODE REVIEW

**Comment by dio** [comment]
@alyssawilk thanks for re-running the job of the ipv6_tests. But it seems it failed again. 

```
fatal: unable to access 'https://boringssl.googlesource.com/boringssl/': Failed to connect to boringssl.googlesource.com port 443: Network is unreachable
 and referenced by '//external:ssl'
```

Anything I can do to make it right? Thanks!

**Comment by alyssawilk** [comment]
Unfortunately, I have no idea.  I'd glanced at this and assumed the hard-coded 127.0.0.1 was the problem (looks like it exists in other tests so probably not) and thankfully checked test logs before (incorrectly) calling it out to you :-P  Last time I saw the boringssl failure doing a re-run resolved the problem which is why I kicked one off for you.

boringssl appears to be up and reachable from here. I wonder if there's network or firewall problems circle-CL side?

**Comment by alyssawilk** [comment]
It totally looks like it is doing other network fetches and just failing on boringssl.  Probably something circleCI side.

I wonder if switching over to https://github.com/google/boringssl would help

**Comment by alyssawilk** [comment]
looks like it's not just you either.

https://circleci.com/gh/envoyproxy/envoy/36395?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Can you try a master merge just to see if that sorts things out?  If not I'll go ping folks at circle and see if they can sort out what's up

**Comment by alyssawilk** [comment]
grr.  Ok, try again past https://github.com/envoyproxy/envoy/pull/2962 ?

**Comment by alyssawilk** [comment]
Oh wait, ipv6 passed this time.  now mac is unhappy.  Oy

**Comment by dio** [comment]
@alyssawilk finally! After 56 minutes and 33 seconds. :tada:

File: source/common/upstream/health_checker_impl.h [hunk 1]
```diff
@@ -226,6 +226,14 @@ class HttpHealthCheckerImpl : public HealthCheckerImplBase {
 
   typedef std::unique_ptr<HttpActiveHealthCheckSession> HttpActiveHealthCheckSessionPtr;
 
+  struct HeaderValueWrapper {
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.h]: Can we share this code in a utility class somehow? I think this was mostly copied from the router or elsewhere? (Same as the config/setting functions above). I suspect we will be here again somewhere else in the future...

File: source/common/upstream/health_checker_impl.h [hunk 2]
```diff
@@ -226,6 +226,14 @@ class HttpHealthCheckerImpl : public HealthCheckerImplBase {
...
+    HeaderValueWrapper(const std::string& value, bool append) : value_(value), append_(append) {}
+
+    const std::string value_;
+    bool append_;
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.h]: nit: const

File: source/common/upstream/health_checker_impl.h (hunk 1 above)
Comment (dio) [reply source/common/upstream/health_checker_impl.h]: > I think this was mostly copied from the router or elsewhere?

Comment (dio) [reply source/common/upstream/health_checker_impl.h]: @mattklein123 I moved it into `common/http/utility` if it is not the appropriate place, please let me know (probably just go ahead to add `HealthCheckUtility` class?). Thanks!

**Comment by mattklein123** [comment]
@dio thanks, in general this looks great. Can you merge master on top of your other PR?

File: source/common/http/utility.cc [hunk 3]
```diff
@@ -300,5 +300,30 @@ void Utility::appendToHeader(HeaderString& header, const std::string& data) {
   header.append(data.c_str(), data.size());
 }
 
+void Utility::configureHeadersToAdd(
```
[the remark this answers is not in the document]
Comment (zuercher) [reply source/common/http/utility.cc]: It's in `Router::HeaderParser` (see `source/common/router/header_parser.h`).

The primary distinction is that code expects a RequestInfo object to support %PROTOCOL% and so on. It doesn't seem like it'd be very hard to set a mode where HeaderValueOption values are treated strictly as strings (no support for %-delimited expressions) and then pass it a default RequestInfo to generate the headers.

[the remark this answers is not in the document]
Comment (dio) [reply source/common/http/utility.cc]: @zuercher I didn't spend enough time to figure that out. Will check it out. Thanks.

[the remark this answers is not in the document]
Comment (dio) [reply source/common/http/utility.cc]: Updated to use `Router::HeaderParser` instead. Thanks again @zuercher.

File: source/common/upstream/health_checker_impl.cc [hunk 4]
```diff
@@ -102,6 +105,7 @@ void HttpHealthCheckerImpl::HttpActiveHealthCheckSession::onInterval() {
       {Http::Headers::get().Path, parent_.path_},
       {Http::Headers::get().UserAgent, Http::Headers::get().UserAgentValues.EnvoyHealthChecker}};
 
+  parent_.request_headers_parser_->evaluateHeaders(request_headers, RequestInfo::RequestInfoImpl());
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: nit: Can you allocate a `static const RequestInfoImpl` and use that? Then we can avoid allocating a temporary each time.

Comment (dio) [reply source/common/upstream/health_checker_impl.cc]: Sure, added in e28f72d. Thanks.

**Pull request merged**